examples/asymcute_mqttsn/main.c

```diff
@@ -0,0 +1,553 @@
+
+#ifndef REQ_CTX_NUMOF
+#define REQ_CTX_NUMOF       (8U)
+#endif
+
+#ifndef SUB_CTX_NUMOF
+#define SUB_CTX_NUMOF       (8U)
+#endif
+
+#ifndef TOPIC_BUF_NUMOF
+#define TOPIC_BUF_NUMOF     (8U + SUB_CTX_NUMOF)
+#endif
+
+/* needed for the `ping6` shell command */
+#define MAIN_QUEUE_SIZE     (8)
+static msg_t _main_msg_queue[MAIN_QUEUE_SIZE];
+
+#define LISTENER_PRIO       (THREAD_PRIORITY_MAIN - 1)
+
+static char listener_stack[ASYMCUTE_LISTENER_STACKSIZE];
+
+static asymcute_con_t _connection;
+static asymcute_req_t _reqs[REQ_CTX_NUMOF];
+static asymcute_sub_t _subscriptions[SUB_CTX_NUMOF];
+static asymcute_topic_t _topics[TOPIC_BUF_NUMOF];
+
+static asymcute_req_t *_get_req_ctx(void)
+{
+    for (unsigned i = 0; i < REQ_CTX_NUMOF; i++) {
+        if (!asymcute_req_in_use(&_reqs[i])) {
+            return &_reqs[i];
+        }
+    }
+    puts("error: no request context available\n");
+    return NULL;
+}
+
+static asymcute_sub_t *_get_sub_ctx(void)
+{
+    for (unsigned i = 0; i < SUB_CTX_NUMOF; i++) {
+        if (!asymcute_sub_active(&_subscriptions[i])) {
+            return &_subscriptions[i];
+        }
+    }
+    return NULL;
+}
+
+static asymcute_sub_t *_find_sub(const char *name)
+{
+    for (unsigned i = 0; i < SUB_CTX_NUMOF; i++) {
+        if (asymcute_sub_active(&_subscriptions[i]) &&
+            strcmp(_subscriptions[i].topic->name, name) == 0) {
+            return &_subscriptions[i];
+        }
+    }
+    return NULL;
+}
+
+static uint16_t _topic_parse_pre(const char *name)
+{
+    if (strncmp(name, "pre_", 4) == 0) {
+        return (uint16_t)atoi(&name[4]);
+    }
+    return 0;
+}
+
+static int _topic_init(asymcute_topic_t *t, const char *name)
+{
+    uint16_t id = _topic_parse_pre(name);
+
+    if (id != 0) {
+        name = NULL;
+    }
+
+    if (asymcute_topic_init(t, name, id) != ASYMCUTE_OK) {
+        return 1;
+    }
+    return 0;
+}
+
+static int _topic_find(asymcute_topic_t *t, const char *name)
+{
+    size_t len = strlen(name);
+    uint16_t id = _topic_parse_pre(name);
+
+    if ((id != 0) || (len == 2)) {
+        if (t) {
+            return _topic_init(t, name);
+        }
+        return 0;
+    }
+
+    /* need to find topic in list of registered ones */
+    for (unsigned i = 0; i < TOPIC_BUF_NUMOF; i++) {
+        if (asymcute_topic_is_reg(&_topics[i]) &&
+            (strncmp(name, _topics[i].name, sizeof(_topics[i].name)) == 0)) {
+            if (t) {
+                memcpy(t, &_topics[i], sizeof(asymcute_topic_t));
+            }
+            return 0;
+        }
+    }
+    return 1;
+}
+
+static void _topics_clear(void)
+{
+    memset(_topics, 0, sizeof(_topics));
+}
+
+static asymcute_topic_t *_topic_get_free(void)
+{
+    for (unsigned i = 0; i < TOPIC_BUF_NUMOF; i++) {
+        if (!asymcute_topic_is_init(&_topics[i])) {
+            return &_topics[i];
+        }
+    }
+    return NULL;
+}
+
+static int _qos_parse(int argc, char **argv, int pos, unsigned *flags)
```

Comment:
Any reason why not just hand over the argument string here instead of the whole array?

Reply:
yes, because I 'outsource' the detection if enough parameters are passed to the shell command to this function. Else I would have to check for `argc > something` always before calling this function.

Reply:
Ah, okay. Missed that part